cpp/src/arrow/dataset/file_csv.cc

```diff
@@ -334,7 +334,10 @@ Status CsvFileWriter::Write(const std::shared_ptr<RecordBatch>& batch) {
   return batch_writer_->WriteRecordBatch(*batch);
 }
 
-Status CsvFileWriter::FinishInternal() { return batch_writer_->Close(); }
+Future<> CsvFileWriter::FinishInternal() {
+  return DeferNotOk(destination_locator_.filesystem->io_context().executor()->Submit(
+      [this]() { return batch_writer_->Close(); }));
+}
```

Comment:
The CSV writer's `Close` method is a no-op so I think it would be safe to just return a finished future here.

```suggestion
Future<> CsvFileWriter::FinishInternal() {
  batch_writer_->Close();
  return Future<>::MakeFinished();
}
```